### DISCUSSION TITLE
Best way for using a solenoid.

### DISCUSSION THREAD

**Discussion by ithinkido** [grblHAL/plugins#15] [Ideas]
I have a solenoid for raising and lowering a pen . What would be the best way for using this in grblHAL ? I have thought about using [M62-M66](http://www.linuxcnc.org/docs/html/gcode/m-code.html#mcode:m62-m65) and AUX out pins , but would there be a better way to do this so that the solenoid could have a pull and hold phase maybe with different PWM values for each ?  

Another related point I am unsure about . What should be done in  the config file for Z- step and Z-Dir pins , as these are not used . Commenting them out does not work, nor defining less than 3 axis.

**Comment by terjeio** [comment]
A plugin that claims the [HAL spindle functions](http://svn.io-engineering.com/grblHAL/html/structspindle__ptrs__t.html) or
1. repurposes spindle on to switch on the solenoid with full  power and after a short delay uses the spindle direction signal to reduce the current - needs appropriate electronics to handle the direction input for current reduction.
2. repurposes spindle PWM by applying 100% (or whatever appropriate) on spindle on and then after a short delay reduces the PWM output to just about the hold current.

This way you can still use M3Sxxx to switch on the solenoid and M5 to switch it off. Or just S1 and S0 after an initial M3...

By saving the current HAL pointers on plugin initialization you can call the original functions so no need to write code for handling the on/direction outputs or the PWM signal.

> What should be done in the config file for Z- step and Z-Dir pins. 

Map them to the same pin? I am not going to modify the code for less than 3 axes - too much work.

**Reply by terjeio** [reply]
Here is an example of how to reduce current via PWM:

[solenoid.zip](https://github.com/grblHAL/plugins/files/9738378/solenoid.zip)

**Comment by ithinkido** [comment]
I have not been able to test this  yet, but I wanted to say thanks very much for the code .

**Comment by ithinkido** [comment]
I got some quick testing in. 
There is a small typo in the code . Easy fix. L52  `set_solenoid_state`  -> `solenoid_set_state`.

I am building on an ESP32 board and have added the `solenoid.c` to  main/CMakeLists.txt, and built and flashed  using the docker method [here](https://github.com/grblHAL/ESP32#using-docker) . There are no errors.
For the sake of tests, I have just used the default  machine map ( [bdring_v4](https://github.com/grblHAL/ESP32/blob/master/main/bdring_v4_map.h) )  which has spindle PWM on GPIO 2 so I have an easy test platform using the onboard LED, on a fresh copy with no other modifications besides  adding the solenoid code. 

Although the LED ( spindle) does switch with M3 / M5, I do not see any reduction in PWM value when activating the spindle ( M3 S1000) after the `SOLENOID_HOLD_DELAY `

For debug , I moved a `hal.stream.write.(..)` line into the init function , and this is not appearing on the startup messages which was surprising  cause the function is started with  `my_plugin_init()` and  should  happen [automatically](https://github.com/grblHAL/Templates/blob/ce286fd09454e25ca3b1a4bc1026decebf6e1a0e/my_plugin/settings/my_plugin.c#L101),  I also used a  big reduction for hold power to make the change very visable.
I am sure I am  still missing something, but I am not sure what. 

[solenoid .c.txt](https://github.com/grblHAL/plugins/files/9749525/solenoid.c.txt)
[CMakeLists.txt](https://github.com/grblHAL/plugins/files/9749520/CMakeLists.txt)

**Reply by terjeio** [reply]
> ...but I am not sure what.

Neither am I - the compiler or linker is not working as it should?

Rename [my_plugin_init()](https://github.com/grblHAL/core/blob/2f42b7226bc9c70b98d612ae1c46576b718153f5/my_plugin.c#L24) in the core to xmy_plugin_init() and try again.

**Reply by terjeio** [reply]
It is a [linker bug (or quirk?)](https://www.esp32.com/viewtopic.php?f=13&t=8418&p=35899).

**Reply by terjeio** [reply]
Checking more I have found that adding the file containing `my_plugin_init()` just before _grbl/my_plugin.c_ in the `SRCS` list makes the linker pick up the correct version:
```
 ...
 grbl/motion_control.c
 solenoid.c
 grbl/my_plugin.c
 ...
```
Another option is to comment out or remove _grbl/my_plugin.c_ from the list.

**Reply by ithinkido** [reply]
Commenting out _grbl/my_plugin.c_ has worked . This sort of linker error was not something I would have ever thought of. Thanks very much for this solution.The plugin is working very well now.
